File: java/stack/src/main/java/com/js/datastructuresalgorithms/stack/JSStack.java
package com.js.datastructuresalgorithms.stack;

import java.util.LinkedList;

public class JSStack<T> {

    //*****************
    // Class Members **
    //*****************
    private LinkedList<T> mList;

    //***************
    // Constructor **
    //***************
    public JSStack() {};

    public int size() { return list().size(); }

    public boolean isEmpty() { return size() == 0; }

    public void push(T item) { list().addFirst(item); }

    public T pop() { return list().getFirst(); }

    public T peek() { return list().peekFirst(); }

    //*******************
    // Private Members **
    //*******************

    private LinkedList<T> list() { return mList == null ? mList = new LinkedList<>() : mList; }
}

File: node/stack/src/types/JSStack.js
class JSStack {

    #mArray = [];
    #mSize = 0;

    size = () => this.#mSize;

    isEmpty = () => this.size() === 0;

    push = value => this.#array()[this.#mSize++] = value;

    pop = () => this.#array()[this.#mSize--];

    peek = () => this.#array()[this.#mSize];

    //*******************
    // Private Members **
    //*******************

    #array = () => this.#mArray;
}

module.exports = JSStack;